controlplane/kubeadm/controllers/remediation.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// reconcileUnhealthyMachines handles KubeadmControlPlane unhealthyMachine reconciliation.
+func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.Context, c *internal.ControlPlane) (ret ctrl.Result, retErr error) {
```

Comment:
If I understand correctly, we enter here even if `MachineOwnerRemediatedCondition` is true and may set it to false below. Do we re-remediate an already remediated machine?

Reply:
This should not happen, because when a machine is remediated it gets deleted, and as discussed in https://github.com/kubernetes-sigs/cluster-api/pull/3830#discussion_r509631563 there are two checks to prevent double remediation.